### PR TITLE
WINTERMUTE: Fix PathUtils

### PR DESCRIPTION
**Pull request by tobiatesan**
This avoids hitting a failing assert when passing strings == "".
E.g. Bickadoodle has a script bug that tries to load an empty filename.

This fixes #6594

### CODE REVIEW

File: engines/wintermute/utils/path_util.cpp [hunk 1]
```diff
@@ -74,6 +74,11 @@ AnsiString PathUtil::getDirectoryName(const AnsiString &path) {
 AnsiString PathUtil::getFileName(const AnsiString &path) {
 	AnsiString newPath = unifySeparators(path);
 	Common::String lastPart = Common::lastPathComponent(newPath, '/');
+
+	if (lastPart.empty()) {
+		return path;
+	}
```
Comment (somaen) [inline engines/wintermute/utils/path_util.cpp]: Technically, the following code should be else-if to this I guess.

On the other hand, does lastPathComponent actually return strings with the separator still in them? (Or could we simply do else { return lastPart; } ?

Comment (lordhoto) [inline engines/wintermute/utils/path_util.cpp]: > On the other hand, does lastPathComponent actually return strings with the separator still in them? (Or could we simply do else { return lastPart; } ?

I'm sure it should not contain any path separator inside the result. So, returning the result of ``lastPathComponent`` should be fine AFAICT. I do not understand the reason for the `if` check below though.

Comment (tobiatesan) [inline engines/wintermute/utils/path_util.cpp]: It is apparently not a WME thing:
https://bitbucket.org/MnemonicWME/wme1/src/62204728c14de49e49d5b6b70f93a00c310f2e0f/src/engine_core/wme_base/utils.cpp?at=default&fileviewer=file-view-default

Apparently, the whole getFilename method can be removed without damage, as in PR #737

Although: I didn't do a _lot_ of testing yet.

Comment (tobiatesan) [inline engines/wintermute/utils/path_util.cpp]: Apparently it can wrap lastPathComponent straight away without much damage.

Beware that only I tested it _a little_.

Comment (tobiatesan) [inline engines/wintermute/utils/path_util.cpp]: Given the context (https://github.com/scummvm/scummvm/commit/afee4aeacccf66f52fefacbb0904049f0ce45227#diff-cb293c2a8565f312766de8a207314d1eL90) I wouldn't be surprised if it was a leftover from... something.

**Comment by tobiatesan** [comment]
Please don't merge just yet, I am in the process of adding a few test cases

**Comment by tobiatesan** [comment]
@lordhoto @somaen I've added a bunch of tests and so far the only change in behaviour is that the segafaulting calls are no longer segfaulting.

Anything weird has yet to happen when actually running it.

I think this is safe.

**Comment by tobiatesan** [comment]
I'm sorry - is there anything blocking this one? (Reminder - this actually causes a crash, it's not just cosmetic)
I guess once this is in along with #714 I could put a few test cases in place and refactor the whole thing in a less redundant fashion, if that's the issue.

File: test/engines/wintermute/path_utils.h [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+				);
+		TS_ASSERT_EQUALS(
+				Wintermute::PathUtil::getDirectoryName(windowsCapPath),
+				Common::String("C:\\SOME\\")
+				);
+
+		/*
+		 * Failing assert in Common::Array
+		 */
+		// TS_ASSERT_EQUALS(
```
Comment (somaen) [inline test/engines/wintermute/path_utils.h]: Why is this here then?

File: test/engines/wintermute/path_utils.h [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+	const Common::String windowsRelativePath;
+	PathUtilTestSuite () :
+		unixPath("/some/file.ext"),
+		unixCapPath("/SOME/FILE.EXT"),
+		windowsPath("C:\\some\\file.ext"),
+		windowsCapPath("C:\\SOME\\FILE.EXT"),
+		emptyString(""),
+		dualExtPath("/some/file.tar.gz"),
+		manyExtPath("/some/file.tar.bz2.gz.zip"),
+		mixedSlashesPath1("C:\\this/IS_REALLY\\weird.exe"),
```
Comment (somaen) [inline test/engines/wintermute/path_utils.h]: Why do we want to allow this?

File: test/engines/wintermute/path_utils.h [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	const Common::String emptyString;
+	const Common::String dualExtPath;
+	const Common::String manyExtPath;
+	const Common::String mixedSlashesPath1;
+	const Common::String mixedSlashesPath2;
+	const Common::String unixRelativePath;
+	const Common::String windowsRelativePath;
+	PathUtilTestSuite () :
+		unixPath("/some/file.ext"),
+		unixCapPath("/SOME/FILE.EXT"),
```
Comment (somaen) [inline test/engines/wintermute/path_utils.h]: How about the case where you have a trailing slash?

File: test/engines/wintermute/path_utils.h (hunk 3 above)
Comment (somaen) [inline test/engines/wintermute/path_utils.h]: Additionally, absolute paths are bad(tm)

File: engines/wintermute/utils/path_util.cpp [hunk 5]
```diff
@@ -73,12 +73,7 @@ Common::String PathUtil::getDirectoryName(const Common::String &path) {
 //////////////////////////////////////////////////////////////////////////
 Common::String PathUtil::getFileName(const Common::String &path) {
 	Common::String newPath = unifySeparators(path);
-	Common::String lastPart = Common::lastPathComponent(newPath, '/');
-	if (lastPart[lastPart.size() - 1 ] != '/') {
-		return lastPart;
-	} else {
-		return path;
-	}
+	return Common::lastPathComponent(newPath, '/');
```
Comment (somaen) [inline engines/wintermute/utils/path_util.cpp]: Does this guarantee that i.e. /foo/bar/ returns ""?

File: test/engines/wintermute/path_utils.h [hunk 6]
```diff
@@ -53,14 +53,10 @@ class PathUtilTestSuite : public CxxTest::TestSuite {
 				Wintermute::PathUtil::getDirectoryName(windowsCapPath),
 				Common::String("C:\\SOME\\")
 				);
-
```
Comment (somaen) [inline test/engines/wintermute/path_utils.h]: Preferably remove these from the preceding commit, and just introduce them here.

File: test/engines/wintermute/path_utils.h (hunk 3 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: Because while absolute paths are the spawn of the devil there are a few games that use them already, by mistake - the CR series seems to be a major offender - and I thought that we might like to test for a defined behavior in that case, just to be sure of what the function is doing in that case.

We might very well want to change that behaviour, of course, but I don't think it has P=0 of happening.

File: test/engines/wintermute/path_utils.h (hunk 4 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: Ah, thanks, good point.

File: test/engines/wintermute/path_utils.h (hunk 2 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: I un-comment it in 76b49c8, after I've fixed the failing behaviour.

I can remove it altogether, I see how it can raise eyebrows but in this particukar circumstances it seemed clearer this way to me.

File: test/engines/wintermute/path_utils.h (hunk 6 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: Okay, thanks

**Comment by sev-** [comment]
What is the status of this PR?

**Comment by tobiatesan** [comment]
I am rebasing right now.

File: test/engines/wintermute/path_utils.h (hunk 4 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: On a second thought, I'm not sure it's so good a point.
I don't think you can put a "directory" in a .DCP, so my guess is that it's probably best if anything with a trailing / is *not* found.

I have added a commit (ac0d476) that takes care of trailing "/", but if you want my opinion I'd guess it would be best *not* to merge that one.

File: test/engines/wintermute/path_utils.h (hunk 6 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: Done

File: test/engines/wintermute/path_utils.h (hunk 4 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: I am an idiot, my commit just reinstates the previous behaviour. Sorry, it's been a few months.

File: engines/wintermute/utils/path_util.cpp (hunk 5 above)
Comment (tobiatesan) [inline engines/wintermute/utils/path_util.cpp]: Now it does. By the way, is this intended behaviour?
I don't think accessing stuff that ends with a / inside .DCPs is defined behaviour, or is it?

File: test/engines/wintermute/path_utils.h (hunk 4 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: Fixed

**Comment by tobiatesan** [comment]
How do you like this?

File: test/engines/wintermute/path_utils.h (hunk 3 above)
Comment (tobiatesan) [inline test/engines/wintermute/path_utils.h]: Notice the big fat disclaimer at the beginning of the file. Please feel free to reword or dictate a different approach. 

However, those *do* pop up from time to time and I am afraid we must handle them on way or the other :|

**Comment by tobiatesan** [comment]
@bluegr I opened a separate PR #874 which includes this commit

**Comment by sev-** [comment]
This is now obsoleted by #874
